File: sparse_ls.py
#
#  Author(s): Pierre Jolivet <pierre.jolivet@enseeiht.fr>
#       Date: 2021-07-11
#
#  Copyright (C) 2021-     Centre National de la Recherche Scientifique
#
#  This script is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
#  If you use this script, you are kindly asked to cite the following article:
#
#  "A Robust Algebraic Domain Decomposition Preconditioner for Sparse Normal Equations",
#  H. Al Daas, P. Jolivet, and J. A. Scott (2021).
#

import petsc4py,sys
petsc4py.init(sys.argv)
from petsc4py import PETSc

A = PETSc.Mat().create(PETSc.COMM_WORLD)
# begin loading phase (section 4.1.1)
viewer = PETSc.Viewer().createBinary(PETSc.Options().getString('-mat_name'))
A.load(viewer)
viewer.destroy()
# end loading phase
x = PETSc.Vec()
b = PETSc.Vec()
c = PETSc.Vec()
ksp = PETSc.KSP().create(PETSc.COMM_WORLD)
ksp.setFromOptions()
pc = ksp.getPC()
normal = False
if pc.getType() == PETSc.PC.Type.QR:
  ksp.setOperators(A,A)
  ksp.setType(PETSc.KSP.Type.PREONLY)
else:
  # begin partitioning phase (section 4.1.1)
  B = A.transposeMatMult(A)
  mpart = PETSc.MatPartitioning().create(PETSc.COMM_WORLD)
  mpart.setFromOptions()
  mpart.setAdjacency(B)
  isp = PETSc.IS()
  mpart.apply(isp)
  rows = PETSc.IS()
  rows = isp.buildTwoSided()
  perm = B.createSubMatrix(rows,rows)
  B = perm
  (m,n) = A.getOwnershipRange()
  isp = PETSc.IS().createStride(n-m,m,1,PETSc.COMM_WORLD)
  perm = A.createSubMatrix(isp,rows)
  A = perm
  for obj in [mpart,isp,rows]:
    obj.destroy()
  # end partitioning phase
  # begin setup phase (section 4.1.2)
  B.setOption(PETSc.Mat.Option.NEW_NONZERO_ALLOCATION_ERR,False)
  B.setOption(PETSc.Mat.Option.SYMMETRIC,True)
  B.shift(B.norm() * 1.0e-10)
  C = PETSc.Mat()
  if not ksp.getType() == PETSc.KSP.Type.LSQR:
    normal = True
    C.createNormal(A)
  ksp.setOperators(C if normal else A,B)
  if pc.getType() == PETSc.PC.Type.HPDDM:
    (m,n) = B.getOwnershipRange()
    cols = PETSc.IS().createStride(n-m,m,1,PETSc.COMM_SELF)
    (m,n) = A.getSize()
    rows.createStride(m,0,1,PETSc.COMM_SELF)
    A.setOption(PETSc.Mat.Option.SUBMAT_SINGLEIS,True)
    Neumann = A.createSubMatrices(rows,cols)
    isp = Neumann[0].findZeroRows()
    B.increaseOverlap(cols)
    Neumann = A.createSubMatrices(rows,cols)
    Neumann[0].zeroRows(isp,0.0)
    aux = PETSc.Mat()
    aux = Neumann[0].transposeMatMult(Neumann[0])
    aux.setOption(PETSc.Mat.Option.NEW_NONZERO_ALLOCATION_ERR,False)
    aux.shift(aux.norm() * 1.0e-8)
    pc.setHPDDMAuxiliaryMat(cols,aux)
    pc.setHPDDMHasNeumannMat(True)
  # end setup phase
# begin solution phase (section 4.1.3)
(x,b) = A.createVecs()
b.setRandom()
c = x.duplicate()
A.multTranspose(b,c)
ksp.solve(c if normal else b,x)
# end solution phase (section 4.1.3)
b.scale(-1.0)
A.multAdd(x,b,b)
A.multTranspose(b,c)
norm = [b.norm(),c.norm()]
if not PETSc.COMM_WORLD.Get_rank():
  print("||A^T(Ax-b)|| / ||Ax-b|| = "+str(norm[1])+" / "+str(norm[0])+" = "+str(norm[1]/norm[0]))

File: sparse_ls.c
/*
   Author(s): Pierre Jolivet <pierre.jolivet@enseeiht.fr>
        Date: 2021-07-11

   Copyright (C) 2021-     Centre National de la Recherche Scientifique

   This script is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

   If you use this script, you are kindly asked to cite the following article:

   "A Robust Algebraic Domain Decomposition Preconditioner for Sparse Normal Equations",
   H. Al Daas, P. Jolivet, and J. A. Scott (2022).
 */

#include <petsc.h>

static char help[] = "Solves a linear least squares problem system using PCHPDDM.\n\n";

int main(int argc,char **args)
{
  Vec         x,b,c;
  Mat         A,B,*Neumann,aux,C = NULL;
  KSP         ksp;
  PC          pc;
  IS          is,rows,cols;
  PetscMPIInt rank;
  PetscReal   norm[2];
  PetscInt    m,n;
  PetscBool   normal = PETSC_FALSE,flg,qr = PETSC_FALSE;
  PetscViewer viewer;
  char        name[PETSC_MAX_PATH_LEN];

  PetscFunctionBeginUser;
  PetscCall(PetscInitialize(&argc,&args,NULL,help));
  PetscCallMPI(MPI_Comm_rank(PETSC_COMM_WORLD,&rank));
  // begin loading phase (section 4.1.1)
  PetscCall(MatCreate(PETSC_COMM_WORLD,&A));
  PetscCall(PetscOptionsGetString(NULL,NULL,"-mat_name",name,sizeof(name),&flg));
  PetscCheck(flg,PETSC_COMM_WORLD,PETSC_ERR_USER,"Must specify matrix name with -mat_name");
  PetscCall(PetscViewerBinaryOpen(PETSC_COMM_WORLD,name,FILE_MODE_READ,&viewer));
  PetscCall(MatLoad(A,viewer));
  PetscCall(PetscViewerDestroy(&viewer));
  // end loading phase
  PetscCall(KSPCreate(PETSC_COMM_WORLD,&ksp));
  PetscCall(KSPSetFromOptions(ksp));
  PetscCall(KSPGetPC(ksp,&pc));
  PetscCall(PetscObjectTypeCompare((PetscObject)pc,PCQR,&flg));
  if (flg) {
    PetscCall(KSPSetOperators(ksp,A,A));
    PetscCall(KSPSetType(ksp,KSPPREONLY));
  } else {
    MatPartitioning mpart;
    Mat             perm;
    // begin partitioning phase (section 4.1.1)
    PetscCall(MatProductCreate(A,A,NULL,&B));
    PetscCall(MatProductSetType(B,MATPRODUCT_AtB));
    PetscCall(MatProductSetFromOptions(B));
    PetscCall(MatProductSymbolic(B));
    PetscCall(MatPartitioningCreate(PETSC_COMM_WORLD,&mpart));
    PetscCall(MatPartitioningSetAdjacency(mpart,B));
    PetscCall(MatPartitioningSetFromOptions(mpart));
    PetscCall(MatPartitioningApply(mpart,&is));
    PetscCall(MatPartitioningDestroy(&mpart));
    PetscCall(ISBuildTwoSided(is,NULL,&rows));
    PetscCall(ISDestroy(&is));
    PetscCall(PetscOptionsGetBool(NULL,NULL,"-pc_use_qr",&qr,NULL));
    if (!qr) PetscCall(MatProductNumeric(B));
    PetscCall(MatCreateSubMatrix(B,rows,rows,MAT_INITIAL_MATRIX,&perm));
    PetscCall(MatHeaderReplace(B,&perm));
    if (!qr) {
      PetscCall(MatNorm(B,NORM_FROBENIUS,norm));
      PetscCall(MatSetOption(B,MAT_NEW_NONZERO_ALLOCATION_ERR,PETSC_FALSE));
      PetscCall(MatSetOption(B,MAT_SYMMETRIC,PETSC_TRUE));
      PetscCall(MatShift(B,*norm * 1.0e-10));
    }
    PetscCall(MatGetOwnershipRange(A,&m,&n));
    PetscCall(ISCreateStride(PETSC_COMM_WORLD,n-m,m,1,&is));
    PetscCall(MatCreateSubMatrix(A,is,rows,MAT_INITIAL_MATRIX,&perm));
    PetscCall(MatHeaderReplace(A,&perm));
    PetscCall(ISDestroy(&is));
    PetscCall(ISDestroy(&rows));
    // end partitioning phase
    // begin setup phase (section 4.1.2)
    PetscCall(PetscObjectTypeCompare((PetscObject)ksp,KSPLSQR,&normal));
    normal = (PetscBool)!normal;
    if (normal || qr) PetscCall(MatCreateNormal(A,&C));
    PetscCall(KSPSetOperators(ksp,normal?C:A,qr?C:B));
    PetscCall(PetscObjectTypeCompare((PetscObject)pc,PCHPDDM,&flg));
    if (flg) {
      flg = PETSC_FALSE;
      PetscCall(PetscOptionsGetBool(NULL,NULL,"-pc_hidden_setup",&flg,NULL));
      if (!flg) {
        PetscCall(MatGetOwnershipRangeColumn(A,&m,&n));
        PetscCall(ISCreateStride(PETSC_COMM_SELF,n-m,m,1,&cols));
        PetscCall(MatGetSize(A,&m,NULL));
        PetscCall(ISCreateStride(PETSC_COMM_SELF,m,0,1,&rows));
        PetscCall(MatSetOption(A,MAT_SUBMAT_SINGLEIS,PETSC_TRUE));
        PetscCall(MatCreateSubMatrices(A,1,&rows,&cols,MAT_INITIAL_MATRIX,&Neumann));
        PetscCall(MatFindZeroRows(*Neumann,&is));
        PetscCall(MatDestroySubMatrices(1,&Neumann));
        PetscCall(MatIncreaseOverlap(B,1,&cols,1));
        PetscCall(MatCreateSubMatrices(A,1,&rows,&cols,MAT_INITIAL_MATRIX,&Neumann));
        PetscCall(ISDestroy(&rows));
        PetscCall(MatSetOption(*Neumann,MAT_NEW_NONZERO_ALLOCATION_ERR,PETSC_FALSE));
        PetscCall(MatZeroRowsIS(*Neumann,is,0.0,NULL,NULL));
        PetscCall(ISDestroy(&is));
        if (!qr) {
          PetscCall(MatTransposeMatMult(*Neumann,*Neumann,MAT_INITIAL_MATRIX,PETSC_DEFAULT,&aux));
          PetscCall(MatNorm(aux,NORM_FROBENIUS,norm));
          PetscCall(MatSetOption(aux,MAT_NEW_NONZERO_ALLOCATION_ERR,PETSC_FALSE));
          PetscCall(MatShift(aux,*norm * 1.0e-8));
        } else PetscCall(MatCreateNormal(Neumann[0],&aux));
        PetscCall(MatDestroySubMatrices(1,&Neumann));
        PetscCall(PCHPDDMSetAuxiliaryMat(pc,cols,aux,NULL,NULL));
        PetscCall(PCHPDDMHasNeumannMat(pc,PETSC_TRUE));
        if (!normal) {
          PCHPDDMCoarseCorrectionType type;
          PetscCall(PCHPDDMGetCoarseCorrectionType(pc,&type));
          if (type == PC_HPDDM_COARSE_CORRECTION_DEFLATED) PetscCall(PCHPDDMSetCoarseCorrectionType(pc,PC_HPDDM_COARSE_CORRECTION_BALANCED));
        }
        PetscCall(ISDestroy(&cols));
        PetscCall(MatDestroy(&aux));
      }
    }
    // end setup phase
    PetscCall(MatDestroy(&B));
    PetscCall(MatDestroy(&C));
  }
  // begin solution phase (section 4.1.3)
  PetscCall(MatCreateVecs(A,&x,&b));
  PetscCall(VecSetRandom(b,NULL));
  PetscCall(VecDuplicate(x,&c));
  PetscCall(MatMultTranspose(A,b,c));
  PetscCall(KSPSolve(ksp,normal?c:b,x));
  // end solution phase
  PetscCall(VecScale(b,-1.0));
  PetscCall(MatMultAdd(A,x,b,b));
  PetscCall(VecNorm(b,NORM_2,norm));
  PetscCall(MatMultTranspose(A,b,c));
  PetscCall(VecNorm(c,NORM_2,norm+1));
  if (!rank) PetscCall(PetscPrintf(PETSC_COMM_SELF,"||A^T(Ax-b)|| / ||Ax-b|| = %f / %f = %f\n",(double)norm[1],(double)norm[0],(double)(norm[1]/norm[0])));
  PetscCall(KSPDestroy(&ksp));
  PetscCall(VecDestroy(&c));
  PetscCall(VecDestroy(&x));
  PetscCall(VecDestroy(&b));
  PetscCall(MatDestroy(&A));
  PetscCall(PetscFinalize());
  return 0;
}
